### PR TITLE
Implement room.Id value object

### PR DESCRIPTION
**Pull request by psrpinto**
Implements `room.Id` value object, with tests.

### CODE REVIEW

File: domain/room/id_test.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	_, err := NewId("!room:matrix.test")
+	if err != nil {
+		t.Errorf("id is valid so there should be no error, got %s", err)
+	}
+}
+
+func TestId(t *testing.T) {
+	id, _ := NewId("!room:matrix.test")
+	if id.Id() != "!room:matrix.test" {
+		t.Error("invalid id")
```
Comment (ashfame) [inline domain/room/id_test.go:43]: `t.Errorf("invalid id, got %s",id.Id())` would be nicer

**Comment by ashfame** [comment]
@psrpinto I was wondering, it's a very subjective thing, but how do you feel about replacing `Id` with `ID` everywhere since its an acronym?

**Comment by psrpinto** [comment]
> I was wondering, it's a very subjective thing, but how do you feel about replacing Id with ID everywhere since its an acronym?

Personally, I prefer to apply camel case to everything, since it removes an extra step (i.e. you don't need to wonder, it's always the same). That's my personal preference.

In this specific case however, ID is not acronym, it's an abbreviation of "Identifier", so I would be against capitalizing it :)

In the end, whatever we decide will fine for me, but I think we should automate it. I took the liberty of opening an issue for that https://github.com/Automattic/neurobot/issues/48

**Comment by ashfame** [comment]
@psrpinto I see. My initial thoughts were: I can work on adapting it, I care for consistency on a subjective matter.

Digging some more, I found a few bits that I want to mention before we decide what to do:
- `golint` tool that I am using also flags this `Id` to be replaced by `ID` and I learnt that it's meant to dictate styles of [Effective Go](https://golang.org/doc/effective_go.html) and [CodeReviewComments](https://golang.org/wiki/CodeReviewComments), which are like gold standard in Go world (Quite opinionated, but that's what Go lang is)
- `golint` is not maintained or actively developed which I think is primarily from a standpoint of "its feature finish" and we don't care for accommodating for everyone's preferences, just a coherent one across the language.
- `staticcheck` is another go lint that's even funded by Google now, but after turning it on, it doesn't even flag anything where I know it should. Maybe it requires configuration, but I like no questions asked opinionated approach of `gofmt` that makes a lot of Go code look similar and hence easy to work with.

That being said, I think I would prefer using `golint` and Go's core team opinionated implementation for the language. What do you think?

**Comment by ashfame** [comment]
Pushed some changes that we can cherry-pick from https://github.com/Automattic/neurobot/compare/room-id...room-id-changes?expand=1

**Pull request merged**